scalding-core/src/main/scala/com/twitter/scalding/Mode.scala

```diff
@@ -63,72 +75,258 @@ object Mode {
     case _ => None
   }
 
-  val CascadingFlowConnectorClassKey = "cascading.flow.connector.class"
-  val CascadingFlowProcessClassKey = "cascading.flow.process.class"
+  val KnownModesMap = Seq(
+    /* The modes are referenced by class *names* to avoid tightly coupling all the jars together. This class, Mode,
+    must be loadable whether one of these is missing or not (it is unlikely all can be loaded at the same time)
+     */
+    "local" -> "com.twitter.scalding.Local",
+    "hadoop2-mr1" -> "com.twitter.scalding.Hadoop2Mr1Mode",
+    "hadoop2-tez" -> "com.twitter.scalding.TezMode", "tez" -> "com.twitter.scalding.TezMode",
+    "hadoop1" -> "com.twitter.scalding.LegacyHadoopMode",
+    "hdfs" -> "com.twitter.scalding.LegacyHadoopMode",
+    "flink" -> "com.twitter.scalding.FlinkMode")
+
+  val KnownTestModesMap = Seq(
+    "local-test" -> "com.twitter.scalding.Test",
+    "hadoop1-test" -> "com.twitter.scalding.HadoopTest",
+    "hadoop2-mr1-test" -> "com.twitter.scalding.Hadoop2Mr1TestMode",
+    "tez-test" -> "com.twitter.scalding.TezTestMode",
+    "flink-test" -> "com.twitter.scalding.FlinkTestMode")
+  // TODO: define hadoop2-mr1 (easy), tez and flink (less easy) classes.
+
+  private def getModeConstructor[M <: Mode](clazzName: String, types: Class[_]*) =
```

Comment:
I wonder if we should just have a contract that a `Mode` needs a constructor that takes exactly one `Args` and exactly one `Configuration`, or even, possibly `Config` which is immutable. Wouldn't that work? I don't want to add too much flexibility if we don't need it.